[algorithms.samdp] add samdp algorithm

lbalicki opened the pull request:
I implemented the the subspace accelerated dominant pole algorithm (https://ieeexplore.ieee.org/document/1717547) which computes the dominant poles of the transfer function of a LTI system. It can be used in the future to add modal MOR methods to pyMOR.

pmli commented:
Hi @lbalicki, thanks for the PR. Adding a test in `src/pymortests/algorithms/` would be good (something simple, maybe a random or a specially constructed example).

pmli commented on src/pymor/algorithms/samdp.py:
```diff
@@ -0,0 +1,462 @@
+        sEmAB = sEmA.apply_inverse(B_defl)
+        Hs = C_defl.dot(sEmAB)
+
+        y_all, _, u_all = spla.svd(Hs)  # use scipy.sparse.linalg.svds(Hs, k=1) instead?
```
Since both `B` and `C` are converted to `VectorArrays`, this implicitly means that the number of inputs and outputs is small, so the dense SVD is fine.
The assumption about the number of inputs and outputs should be documented (or maybe `B` and `C` should be given as `VectorArrays`, as in `solve_lyap_lrcf`).

pmli commented on src/pymor/algorithms/samdp.py:
```diff
@@ -0,0 +1,462 @@
+
+        X.append(x)
+        V.append(v)
+        V = gram_schmidt(V, atol=0, rtol=0, offset=len(V) - 1, copy=False)
```
```suggestion
        gram_schmidt(V, atol=0, rtol=0, offset=len(V) - 1, copy=False)
```
Similarly below with other calls to `gram_schmidt`.

pmli commented on src/pymortests/algorithms/samdp.py:
```diff
@@ -0,0 +1,90 @@
+# This file is part of the pyMOR project (http://www.pymor.org).
+# Copyright 2013-2019 pyMOR developers and contributors. All rights reserved.
```
```suggestion
# Copyright 2013-2020 pyMOR developers and contributors. All rights reserved.
```

pmli commented on docs/source/bibliography.rst:
```diff
@@ -107,6 +107,11 @@ Bibliography
+.. [RM06] J. Rommes and N. Martins,
+          Efficient computation of multivariable transfer function dominant poles
+          using subspace acceleration.
+          IEEE Trans. on Power Systems 21 (4), pp. 1471-1483, 2006
```
`IEEE Trans. Power Syst.` is the [official abbrevation](http://ieeeauthorcenter.ieee.org/wp-content/uploads/Journal-Titles-and-Abbreviations.pdf).

lbalicki commented:
I made the suggested changes and in order to be more consistent with the other algorithms I decided to change the `B` and `C` parameters to `VectorArrays`.

pmli commented on src/pymor/algorithms/samdp.py:
```diff
@@ -0,0 +1,459 @@
+                        st = shifts[shift_nr]
+                        shift_nr = shift_nr + 1
+            elif k >= krestart:
+                logger.info(f'Perform restart...')
```
Unnecessary f-string.

pmli commented on src/pymor/algorithms/samdp.py:
```diff
@@ -0,0 +1,459 @@
+            rightev = rightev[idx]
+            leftev = leftev[idx]
+            if nr_converged < nwanted:
+                logger.warning(f'The specified number of poles could not be computed.')
```
Unnecessary f-string.

pmli commented on src/pymor/algorithms/samdp.py:
```diff
@@ -0,0 +1,459 @@
+        Residual of the computed triplet.
+    """
+
+    iter = 0
```
`iter` is a Python built-in, it's better not to redefine it.

pmli commented on src/pymor/algorithms/samdp.py:
```diff
@@ -0,0 +1,459 @@
+        return x, y, theta, init_res
+
+
+def select_max_eig(H, G, X, V, B, C, E, which):
```
Argument `E` is unused in this function.

pmli commented on src/pymor/algorithms/samdp.py:
```diff
@@ -0,0 +1,459 @@
+    for i in range(len(H)):
+        V[i].scal(1 / V[i].norm())
+        X[i].scal(1 / X[i].norm())
+        residue = np.append(residue, spla.norm(C.dot(X[i]) @ V[i].dot(B), 2))
```
Can this be done without a for loop? Also, the 2-norm of an outer product of two vectors should be equal to the product of their norms (`||x y^T|| = ||x|| ||y||`).

pmli commented on src/pymor/algorithms/samdp.py:
```diff
@@ -0,0 +1,448 @@
+
+    V.scal(1 / V.norm())
+    X.scal(1 / X.norm())
+    residue = spla.norm(C.dot(X), 2, 0) * spla.norm(V.dot(B), 2, 1)
```
The 2-norm is the default for vectors.
```suggestion
    residue = spla.norm(C.dot(X), axis=0) * spla.norm(V.dot(B), axis=1)
```

pmli commented on src/pymor/algorithms/samdp.py:
```diff
@@ -0,0 +1,448 @@
+    Parameters
+    ----------
+    H
+        The |Numpy array| H from the samdp algorithm.
```
The algorithm name should be uppercase. Similar below.
```suggestion
        The |Numpy array| H from the SAMDP algorithm.
```

pmli commented on src/pymor/algorithms/samdp.py:
```diff
@@ -0,0 +1,448 @@
+        y = y_all[:, 0]
+
+        x = sEmAB.lincomb(u)
+        v = sEmA.apply_inverse_adjoint((C_defl.lincomb(y.T)))
```
```suggestion
        v = sEmA.apply_inverse_adjoint(C_defl.lincomb(y.T))
```

lbalicki commented:
> Some more minor remarks below.
> 
> I tried running the code with the building model from [SLICOT](http://slicot.org/20-site/126-benchmark-examples-for-model-reduction):
> 
> ```python
> from pymor.basic import *
> from pymor.algorithms.samdp import samdp
> build = LTIModel.from_mat_file('build.mat')
> dp = samdp(build.A, build.E, build.B.as_range_array(), build.C.as_source_array(), 2)
> ```
> 
> but I get NaNs appearing.

I had set the default initial shift to zero, which seems to cause problems in some cases. Now, if no initial shifts are provided, a random complex shift is used. For me the model from [SLICOT](http://slicot.org/20-site/126-benchmark-examples-for-model-reduction) works this way.

pmli commented on src/pymor/algorithms/samdp.py:
```diff
@@ -0,0 +1,448 @@
+    shifts = init_shifts
+
+    while nrestart < maxrestart:
+        k = k + 1
```
```suggestion
        k += 1
```

pmli commented on src/pymor/algorithms/samdp.py:
```diff
@@ -0,0 +1,448 @@
+                Q[-1].scal(1 / nqqt)
+                Qs[-1].scal(1 / nqqt)
+
+                nr_converged = nr_converged + 1
```
```suggestion
                nr_converged += 1
```

pmli commented on src/pymor/algorithms/samdp.py:
```diff
@@ -0,0 +1,448 @@
+                    gram_schmidt(X, atol=0, rtol=0, copy=False)
+
+                B_defl = B_defl - E.apply(Q[-1].lincomb(Qt[-1].dot(B_defl).T))
+                C_defl = C_defl - E.apply_adjoint(Qt[-1].lincomb(Q[-1].dot(C_defl).T))
```
```suggestion
                B_defl -= E.apply(Q[-1].lincomb(Qt[-1].dot(B_defl).T))
                C_defl -= E.apply_adjoint(Qt[-1].lincomb(Q[-1].dot(C_defl).T))
```

pmli commented on src/pymor/algorithms/samdp.py:
```diff
@@ -0,0 +1,448 @@
+                B_defl = B_defl - E.apply(Q[-1].lincomb(Qt[-1].dot(B_defl).T))
+                C_defl = C_defl - E.apply_adjoint(Qt[-1].lincomb(Q[-1].dot(C_defl).T))
+
+                k = k - 1
```
```suggestion
                k -= 1
```

pmli commented on src/pymor/algorithms/samdp.py:
```diff
@@ -0,0 +1,448 @@
+                            gram_schmidt(X, atol=0, rtol=0, copy=False)
+
+                            B_defl = B_defl - E.apply(Q[-1].lincomb(Qt[-1].dot(B_defl).T))
+                            C_defl = C_defl - E.apply_adjoint(Qt[-1].lincomb(Q[-1].dot(C_defl).T))
```
```suggestion
                            B_defl -= E.apply(Q[-1].lincomb(Qt[-1].dot(B_defl).T))
                            C_defl -= E.apply_adjoint(Qt[-1].lincomb(Q[-1].dot(C_defl).T))
```

pmli commented on src/pymor/algorithms/samdp.py:
```diff
@@ -0,0 +1,448 @@
+
+                    if shift_nr < nr_shifts:
+                        st = shifts[shift_nr]
+                        shift_nr = shift_nr + 1
```
```suggestion
                        shift_nr += 1
```

pmli commented on src/pymor/algorithms/samdp.py:
```diff
@@ -0,0 +1,448 @@
+                G = V.dot(E.apply(X))
+                AX = A.apply(X)
+                H = V.dot(AX)
+                nrestart = nrestart + 1
```
```suggestion
                nrestart += 1
```

pmli commented on src/pymor/algorithms/samdp.py:
```diff
@@ -0,0 +1,448 @@
+    i = 0
+    nrq = 1
+    while nrq > tol and i < maxiter:
+        i = i + 1
```
```suggestion
        i += 1
```

lbalicki commented:
It seems like there was still an issue when the eigenvectors of the dominant poles were complex. With the changes I have made, the CDplayer model from SLICOT works even with large values for `nwanted`.

pmli commented:
Ok, great, CDplayer is working nicely, and a few other benchmarks also perform better (e.g., random). I have a few more comments now:

1. I noticed that multiple runs can give different results, at least in terms of convergence. Therefore, it would be good to give the user the option to pick the seed for `np.random.uniform`, or maybe just give a suggestion in the documentation to try setting a different initial shift if there is no convergence.
2. There are two more methods here (`twosided_rqi` and `select_max_eig`) that might be better as "private" methods (i.e., renaming to `_twosided_rqi` and `_select_max_eig`) if users are not expected to use them directly.
3. I tried `samdp` on [filter2D](https://morwiki.mpi-magdeburg.mpg.de/morwiki/index.php/Tunable_Optical_Filter) and [rail1357](https://morwiki.mpi-magdeburg.mpg.de/morwiki/index.php/Steel_Profile) benchmarks from the Oberwolfach collection. For larger `nwanted` (e.g., 20), it returns poles with non-zero imaginary parts (although, relatively small), even though both models have only real poles. Do you see a way to remedy this?

lbalicki commented:
> Ok, great, CDplayer is working nicely, and a few other benchmarks also perform better (e.g., random). I have a few more comments now:
> 
>     1. I noticed that multiple runs can give different results, at least in terms of convergence. Therefore, it would be good to give the user the option to pick the seed for `np.random.uniform`, or maybe just give a suggestion in the documentation to try setting a different initial shift if there is no convergence.
> 
>     2. There are two more methods here (`twosided_rqi` and `select_max_eig`) that might be better as "private" methods (i.e., renaming to `_twosided_rqi` and `_select_max_eig`) if users are not expected to use them directly.
> 
>     3. I tried `samdp` on [filter2D](https://morwiki.mpi-magdeburg.mpg.de/morwiki/index.php/Tunable_Optical_Filter) and [rail1357](https://morwiki.mpi-magdeburg.mpg.de/morwiki/index.php/Steel_Profile) benchmarks from the Oberwolfach collection. For larger `nwanted` (e.g., 20), it returns poles with non-zero imaginary parts (although, relatively small), even though both models have only real poles. Do you see a way to remedy this?

Regarding 1. and 2. I just added a `seed=0` option and I changed the name of the functions.

For 3. I changed the default tolerances to `tol=1e-11` and `imagtol=1e-7` which makes the poles of both examples have zero imaginary part. In general the user might have to adjust these tolerances to get accurate results. I also tried to use relative error tolerances in some parts of the algorithm, but that did not lead to better results for these models either.

pmli commented:
> For 3. I changed the default tolerances to `tol=1e-11` and `imagtol=1e-7` which makes the poles of both examples have zero imaginary part. In general the user might have to adjust these tolerances to get accurate results. I also tried to use relative error tolerances in some parts of the algorithm, but that did not lead to better results for these models either.

Ok, interesting. I guess the algorithm is really sensitive to the choices of the tolerances. I tried the new tolerances for the 'random' example in the SLICOT benchmark and it looks like `samdp` can return less than `nwanted` poles. E.g., if `nwanted` is 10, for `tol=1e-11` it returns 4, while for `tol=1e-10` it returns 20. @lbalicki Do you get the same? Is this a feature or a bug?

Also, it would be good to return the test for `which='LM'`. I see that one test would fail. Maybe the assert condition can be relaxed or the tolerances adjusted.

lbalicki commented:
It seems like the problem with the `tol` option is that it can happen that the `_twosided_rqi` function is not able to converge (i.e., the residual will not drop below `tol` no matter how many iterations are performed). In this case, restarts are performed until the maximum amount of restarts is reached and `samdp` can not return the desired amount of poles.

My suggestion would be to print a warning if  `_twosided_rqi` does not converge to the desired tolerance, but continue with the computed poles anyway. This works well for the 'random' `SLICOT` benchmark. Does this make sense?

For the tests I would maybe change `assert np.average(val) < np.average(dom_val)` to `assert np.median(val) < np.median(dom_val)` instead of relaxing the condition.

pmli commented:
> My suggestion would be to print a warning if `_twosided_rqi` does not converge to the desired tolerance, but continue with the computed poles anyway. This works well for the 'random' `SLICOT` benchmark. Does this make sense?

It makes sense to me.

pmli commented:
Now the rail1357 example (for `nwanted` equal to 20) looks wrong. It seems `samdp` does not return poles in conjugate pairs.

lbalicki commented:
I made a few changes which make the algorithm more stable. Also, dealing with complex shifts works better now.

pmli commented:
> Now the rail1357 example (for `nwanted` equal to 20) looks wrong. It seems `samdp` does not return poles in conjugate pairs.

It looks like there are now similar issues for the transmission line model from SLICOT. For `nwanted=50`, it returns only real poles, and with `imagtol=1e-8` it doesn't return conjugate pairs.

lbalicki commented:
Comparing with the poles computed by `scipy.linalg.eig` it looks like it is correct that the poles are real for `nwanted=50` in that example. Nonetheless, I fixed the issue with the conjugate pairs and made some other improvements as well.

pmli commented:
These are the poles I get for tline model, colored by dominance (log10 of norm(residual)/norm(pole.real)):
![tline_poles](https://user-images.githubusercontent.com/1268421/78988578-9dd7c900-7b31-11ea-82d6-bcc65bc1cdd5.png)
This is the result of current `samdp` for `nwanted=50`:
![tline_samdp_50](https://user-images.githubusercontent.com/1268421/78988683-f14a1700-7b31-11ea-8848-7052f3fd6f81.png)
and using an older version (39e0a03e):
![tline_samdp_50_old](https://user-images.githubusercontent.com/1268421/78988707-feff9c80-7b31-11ea-97e5-278c0f33b086.png)
The older version looks better to me...

The results were generated using [tline_samdp.zip](https://github.com/pymor/pymor/files/4461391/tline_samdp.zip)

lbalicki commented:
It seems that this happened because I switched to relative tolerances for the residuals. After some testing it seems that absolute tolerances work better for almost all examples.

One example that does not work better is the [filter2D model](https://morwiki.mpi-magdeburg.mpg.de/morwiki/index.php/Tunable_Optical_Filter). I get a complex pole for `tol=1e-10` and `nwanted=50` with the newest version. The model works fine for `tol=1e-15` but then other models fail (e.g. the building model from [SLICOT](http://slicot.org/20-site/126-benchmark-examples-for-model-reduction)). Overall, it seems to be difficult to balance the `tol` option such that all examples yield good results and regarding this, the implementation [here](https://sites.google.com/site/rommes/software) behaves similarly. Is this behaviour acceptable?

drittelhacker commented:
> I can confirm that `samdp` works well for all the SLICOT benchmarks I tried.
> 
> From what I've heard from @drittelhacker, sensitivity to tolerances is somewhat acceptable with this algorithm.

I would not exactly say "acceptable" but "to be expected"

pmli commented on AUTHORS.md:
```diff
@@ -13,6 +13,9 @@
 * Tim Keil, tim.keil@uni-muenster.de
     * second order derivatives for parameters
 
+* Linus Balicki, linus.balicki@ovgu.de
+    * subspace accelerated dominant pole algorithm in algorithms.samdp
+
```
Looking below, it seems alphabetic ordering is preferred. (Also, pushing another commit might help with the testing.)

pmli commented:
It's merged. Thank you, @lbalicki.
